File: app/src/main/java/app/deadmc/devnetworktool/fragments/rest/RestFragment.java
package app.deadmc.devnetworktool.fragments.rest;

import android.os.Bundle;
import android.os.StrictMode;
import android.support.design.widget.TabLayout;
import android.support.v4.view.ViewPager;
import android.util.Log;
import android.util.SparseArray;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import java.util.HashMap;

import app.deadmc.devnetworktool.R;
import app.deadmc.devnetworktool.adapters.RestPagerAdapter;
import app.deadmc.devnetworktool.constants.DevConsts;
import app.deadmc.devnetworktool.fragments.ObserverFragment;
import app.deadmc.devnetworktool.fragments.ParentFragment;
import app.deadmc.devnetworktool.modules.ResponseDev;
import app.deadmc.devnetworktool.modules.RestRequestHistory;
import app.deadmc.devnetworktool.observables.OkHttpObservable;
import app.deadmc.devnetworktool.singletons.SharedData;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;
import okhttp3.Response;

/**
 * Created by Feren on 02.03.2017.
 */
public class RestFragment extends ObserverFragment {

    private String currentUrl;


    private RestPagerAdapter restPagerAdapter;
    private ViewPager viewPager;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        myFragmentView = inflater.inflate(R.layout.fragment_rest, container, false);
        restoreState(savedInstanceState);
        initElements();
        return myFragmentView;
    }

    @Override
    public void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        //Log.e("onSaveInstanceState", "save");
        outState.putString("currentUrl", currentUrl);
    }

    private void restoreState(Bundle bundle) {
        if (bundle==null)
            return;
        if (!bundle.containsKey("currentUrl"))
            return;
        //Log.e("restoreState","restore is ok");
        currentUrl = bundle.getString("currentUrl");

    }

    @Override
    public void initElements() {
        typeOfFragment = DevConsts.REST;
        super.initElements();
        createObserver(Response.class);
        initViewPager();
    }

    public void initViewPager() {
        viewPager = (ViewPager) myFragmentView.findViewById(R.id.viewpager);
        restPagerAdapter = new RestPagerAdapter(getFragmentManager(),getActivity(),currentUrl,this);
        viewPager.setAdapter(restPagerAdapter);
        viewPager.setOffscreenPageLimit(5);
        TabLayout tabLayout = (TabLayout) myFragmentView.findViewById(R.id.sliding_tabs);
        tabLayout.setupWithViewPager(viewPager);
        viewPager.addOnPageChangeListener(new ViewPager.OnPageChangeListener() {
            public void onPageScrollStateChanged(int state) {}
            public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {}

            public void onPageSelected(int position) {
                if (position == 2) {
                    SparseArray<ParentFragment> parentFragmentSparseArray = restPagerAdapter.getRegisteredFragment();
                    for(int i = 0; i < parentFragmentSparseArray.size(); i++) {
                        int key = parentFragmentSparseArray.keyAt(i);
                        ParentFragment parentFragment = parentFragmentSparseArray.get(key);
                        if (parentFragment instanceof  RestHistoryFragment) {
                            RestHistoryFragment restHistoryFragment = (RestHistoryFragment) parentFragment;
                            restHistoryFragment.initElements();
                        }
                    }
                }

            }
        });
    }

    public void sendRequest(final String url, final String requestMethod, final HashMap<String,String> headers, final HashMap<String,String> body) {
        Log.e("sendRequest","url: "+url);

        StrictMode.ThreadPolicy policy = new StrictMode.ThreadPolicy.Builder().permitAll().build();
        StrictMode.setThreadPolicy(policy);
        observable = OkHttpObservable.getObservable(url,requestMethod,headers,body);
        observable.subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .unsubscribeOn(Schedulers.io())
                .subscribe(observer);

    }

    public void setCurrentUrl(String currentUrl) {
        this.currentUrl = currentUrl;
    }

    private void notifyAllNestedFragments(ResponseDev responseDev) {
        SharedData.getInstance().setResponseDev(responseDev);
        SparseArray<ParentFragment> parentFragmentSparseArray = restPagerAdapter.getRegisteredFragment();
        for(int i = 0; i < parentFragmentSparseArray.size(); i++) {
            int key = parentFragmentSparseArray.keyAt(i);
            ParentFragment parentFragment = parentFragmentSparseArray.get(key);
            if (parentFragment instanceof  ResponseRestFragment) {
                try {
                    ResponseRestFragment responseRestFragment = (ResponseRestFragment) parentFragment;
                    responseRestFragment.setResponse(responseDev);
                } catch (Exception e) {}
            }
        }
    }

    public void loadRestHistory(RestRequestHistory restRequestHistory) {
            SparseArray<ParentFragment> parentFragmentSparseArray = restPagerAdapter.getRegisteredFragment();
            for(int i = 0; i < parentFragmentSparseArray.size(); i++) {
                int key = parentFragmentSparseArray.keyAt(i);
                ParentFragment parentFragment = parentFragmentSparseArray.get(key);
                if (parentFragment instanceof  RequestRestFragment) {
                    RequestRestFragment requestRestFragment = (RequestRestFragment) parentFragment;
                    requestRestFragment.setRequestHistory(restRequestHistory);
                }
            }
        viewPager.setCurrentItem(0, true);
    }

    public void setRestFragment() {
        SparseArray<ParentFragment> parentFragmentSparseArray = restPagerAdapter.getRegisteredFragment();
        for(int i = 0; i < parentFragmentSparseArray.size(); i++) {
            int key = parentFragmentSparseArray.keyAt(i);
            ParentFragment parentFragment = parentFragmentSparseArray.get(key);
            if (parentFragment instanceof  RequestRestFragment) {
                RequestRestFragment requestRestFragment = (RequestRestFragment) parentFragment;
                requestRestFragment.setRestFragment(RestFragment.this);
            }

            if (parentFragment instanceof  RestHistoryFragment) {
                RestHistoryFragment requestRestFragment = (RestHistoryFragment) parentFragment;
                requestRestFragment.setRestFragment(RestFragment.this);
            }
        }
    }

    @Override
    protected void handleResult(Object value) {

        ResponseDev responseDev = (ResponseDev) value;
        unsubscribe();
        notifyAllNestedFragments(responseDev);

    }

}

File: app/src/main/java/app/deadmc/devnetworktool/fragments/rest/RestFragment.kt
package app.deadmc.devnetworktool.fragments.rest

import android.os.Bundle
import android.os.StrictMode
import android.support.design.widget.TabLayout
import android.support.v4.view.ViewPager
import android.util.Log
import android.util.SparseArray
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup

import java.util.HashMap

import app.deadmc.devnetworktool.R
import app.deadmc.devnetworktool.adapters.RestPagerAdapter
import app.deadmc.devnetworktool.constants.DevConsts
import app.deadmc.devnetworktool.fragments.BaseFragment
import app.deadmc.devnetworktool.fragments.ObserverFragment
import app.deadmc.devnetworktool.fragments.ParentFragment
import app.deadmc.devnetworktool.modules.ResponseDev
import app.deadmc.devnetworktool.modules.RestRequestHistory
import app.deadmc.devnetworktool.observables.OkHttpObservable
import app.deadmc.devnetworktool.presenters.RestPresenter
import app.deadmc.devnetworktool.singletons.SharedData
import com.arellomobile.mvp.presenter.InjectPresenter
import com.arellomobile.mvp.presenter.PresenterType
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import kotlinx.android.synthetic.main.fragment_rest.view.*
import okhttp3.Response

class RestFragment : BaseFragment() {
    @InjectPresenter(type = PresenterType.LOCAL)
    lateinit var restPresenter:RestPresenter
    private lateinit var restPagerAdapter: RestPagerAdapter

    override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        myFragmentView = inflater!!.inflate(R.layout.fragment_rest, container, false)
        restoreState(savedInstanceState)
        initElements()
        return myFragmentView
    }

    private fun restoreState(bundle: Bundle?) {
        if (bundle == null)
            return
        if (!bundle.containsKey("currentUrl"))
            return
        restPresenter.currentUrl = bundle.getString("currentUrl")

    }

    fun initElements() {
        //createObserver(Response::class.java)
        initViewPager()
    }

    fun initViewPager() {
        restPagerAdapter = RestPagerAdapter(fragmentManager, activity, this)
        myFragmentView.viewPager.adapter = restPagerAdapter
       myFragmentView.viewPager.offscreenPageLimit = 5
        myFragmentView.tabLayout.setupWithViewPager(myFragmentView.viewPager)
       myFragmentView.viewPager.addOnPageChangeListener(object : ViewPager.OnPageChangeListener {
            override fun onPageScrollStateChanged(state: Int) {}
            override fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int) {}

            override fun onPageSelected(position: Int) {
                if (position == 2) {
                    val parentFragmentSparseArray = restPagerAdapter.registeredFragment
                    for (i in 0 until parentFragmentSparseArray.size()) {
                        val key = parentFragmentSparseArray.keyAt(i)
                        val parentFragment = parentFragmentSparseArray.get(key)
                        (parentFragment as? RestHistoryFragment)?.initElements()
                    }
                }

            }
        })
    }



    private fun notifyAllNestedFragments(responseDev: ResponseDev) {
        SharedData.getInstance().responseDev = responseDev
        val parentFragmentSparseArray = restPagerAdapter!!.registeredFragment
        for (i in 0 until parentFragmentSparseArray.size()) {
            val key = parentFragmentSparseArray.keyAt(i)
            val parentFragment = parentFragmentSparseArray.get(key)
            if (parentFragment is ResponseRestFragment) {
                try {
                    parentFragment.setResponse(responseDev)
                } catch (e: Exception) {
                }

            }
        }
    }

    fun loadRestHistory(restRequestHistory: RestRequestHistory) {
        val parentFragmentSparseArray = restPagerAdapter!!.registeredFragment
        for (i in 0 until parentFragmentSparseArray.size()) {
            val key = parentFragmentSparseArray.keyAt(i)
            val parentFragment = parentFragmentSparseArray.get(key)
            /*
                if (parentFragment instanceof  RequestRestFragment) {
                    RequestRestFragment requestRestFragment = (RequestRestFragment) parentFragment;
                    requestRestFragment.setRequestHistory(restRequestHistory);
                }
                */
        }
       myFragmentView.viewPager.setCurrentItem(0, true)
    }

    fun setRestFragment() {
        val parentFragmentSparseArray = restPagerAdapter!!.registeredFragment
        for (i in 0 until parentFragmentSparseArray.size()) {
            val key = parentFragmentSparseArray.keyAt(i)
            val parentFragment = parentFragmentSparseArray.get(key)
            /*
            if (parentFragment instanceof  RequestRestFragment) {
                RequestRestFragment requestRestFragment = (RequestRestFragment) parentFragment;
                requestRestFragment.setRestFragment(RestFragment.this);
            }

            if (parentFragment instanceof  RestHistoryFragment) {
                RestHistoryFragment requestRestFragment = (RestHistoryFragment) parentFragment;
                requestRestFragment.setRestFragment(RestFragment.this);
            }
            */
        }
    }

    override fun handleResult(value: Any) {

        val responseDev = value as ResponseDev
        unsubscribe()
        notifyAllNestedFragments(responseDev)

    }

}
